Feature/socket wrapper

shulard opened the pull request:
Hello !

Here the PR about the implementation of hoaproject/Socket#27 in the websocket project.

I tried to make it clean :
- Must use `WebsocketSocket` instead of just `Socket` to avoid `Socket` name reusing (due to `Hoa\Socket\Socket` flexion) ;
- Retrieve `endPoint` from the socket if it is not given to `Websocket\Client` constructor ;
- Add the factories to `Websocket\Connection`.

But I have a problem about factories loading... They are defined in the Websocket lib so if I construct a `Socket\Client` before using `Websocket\Client` the factories are not yet loaded...

Examples :

``` php
<?php
$connection = new Socket\Client('wss://0.0.0.0/a/path');
// Here the transport wrappers are not yet registered because classes are loaded
// only at first use
$client = new Websocket\Client($connection);
// They are defined here but too late, I already get a "wss transport does not exists" exception
```

``` php
<?php
// That way it works because PHP parser encounter the `Websocket\Client` token
// before the `Socket\Client`
$client = new Websocket\Client(
    new Socket\Client('wss://0.0.0.0/a/path')
);
```

For me there is 3 approachs to use here :
- Create an extension of `Socket\Client` and use it instead of the root class to enable and use the wrappers properly ;
- Change the register approach (I haven't idea here) ;
- Create a factory for the Websocket\Client which handle `Socket\Client` creation.

I think it's not really valid because sometimes the code will works, sometimes not because of instruction order...

I haven't added tests for the moment because we need to validate the approach before...

shulard committed:
Add the Socket extension.

A WebSocket socket must have some specific behaviours.

shulard committed:
Add socket wrapper method for wss and ws scheme.

Register it directly in the Connection object.

Hywan commented on Client.php:
```diff
@@ -84,11 +84,16 @@ class Client extends Connection
         parent::__construct($client);
-        $this->setEndPoint($endPoint);
+
+        $this->setEndPoint(is_null($endPoint)?'/':$endPoint);
```
Correct. We prefer `null ==` instead of `is_null` but that's a detail.
Changing for `null` is correct, bravo.

shulard replied:
Ok, I'll update to `null ==`

Hywan commented on Client.php:
```diff
@@ -84,11 +84,16 @@ class Client extends Connection
+        if( is_null($endPoint) &&
+            $client->getSocket() instanceof WebsocketSocket ) {
+            $this->setEndPoint($client->getSocket()->getEndPoint());
+        }
```
We are testing `$endPoint` twice. Maybe the following flow is more appropriated:
```php
if (null === $endPoint) {
    $endPoint = '/';

    if ($client->getSocket() instanceof …) {
        $endPoint = $client->getSocket()->getEndPoint();
    }

    $this->setEndPoint($endPoint);
}
```
I might be wrong. Thoughts?

shulard replied:
No, you are right :smile: I'll update

Hywan commented on Client.php:
```diff
@@ -152,6 +157,12 @@ protected function doHandshake()
         $connection = $this->getConnection();
         $connection->connect();
+
+        if( $connection->getSocket() instanceof WebsocketSocket &&
```
Correct, bravo!

Hywan commented on Connection.php:
```diff
@@ -555,3 +555,26 @@ public function close($code = self::CLOSE_NORMAL, $reason = null)
 }
+
+/**
+ * Register socket wrappers
```
What do you think about moving this code as a static method on the `Hoa\Websocket\Connection` class? Would it make sense?

shulard replied:
Maybe, but the flow is not better with a static method... If we moved here we always need to think calling the static before start working...

Hywan replied:
Yup, correct, but we can unit test the static method :wink:.

shulard replied:
Ok, it's still better than floating code... I create a registerWrappers method.

Hywan replied:
Or just `getSocketWrapper`, because the method is not responsible to register it, just to provide the “factory”.

Hywan commented on Connection.php:
```diff
@@ -555,3 +555,26 @@ public function close($code = self::CLOSE_NORMAL, $reason = null)
+$factory = function($socketUri) {
+    $parsed = parse_url($socketUri);
+
+    $secure = 'wss' === $parsed['scheme'];
```
What if `scheme` index is undefined?

Hywan commented on Connection.php:
```diff
@@ -555,3 +555,26 @@ public function close($code = self::CLOSE_NORMAL, $reason = null)
+    }
+
+    return new WebsocketSocket(
+        'tcp://' . $parsed['host'] . ':' . $port,
```
What if `port` index is undefined?
What to do with `user` and `pass` indexes too?

shulard replied:
About the `user:pass` I don't know which must be done... Are the sockets able to handle `Basic Auth` ?

They are not handled in the `Hoa\Socket` too... Is the Websocket a specific case for authentication ?

Hywan replied:
Hmm, good point. I guess we can skip it for now.

Hywan commented on Connection.php:
```diff
@@ -555,3 +555,26 @@ public function close($code = self::CLOSE_NORMAL, $reason = null)
+        isset($parsed['path'])?$parsed['path']:'/'
+    );
+};
+\Hoa\Socket\Transport::registerWrapper('ws', $factory);
```
```php
use Hoa\Socket;

…

Socket\Transport::registerWrapper('w…
```

Hywan commented on WebsocketSocket.php:
```diff
@@ -0,0 +1,102 @@
+ * @copyright  Copyright © 2007-2015 Hoa community
+ * @license    New BSD License
+ */
+class WebsocketSocket extends Socket
```
Why not `Socket`? With `use` you can easily write `use Hoa\Socket as HoaSocket;`.

shulard replied:
Ok, if the `use ... as HoaSocket` is right for you...

Basically, I started with `use ... as BaseSocket` which is ugly and I moved to `WebsocketSocket` (which is ugly too :smile:)... Consistency is always the answer ;)

Hywan commented on WebsocketSocket.php:
```diff
@@ -0,0 +1,102 @@
+     */
+    public function __construct($uri, $secure = false, $endPoint = '/')
+    {
+        parent::__construct($uri);
```
I wonder if the factory can just be moved here. Not really the factory, but the `ws://` to `tcp://` translation.
Thoughts?

shulard replied:
Maybe it can...

I submitted something similar here :

* https://github.com/shulard/Websocket/blob/feature/application-transport-layer/Socket.php#L67-L83
* https://github.com/shulard/Websocket/blob/feature/application-transport-layer/Connection/Connection.php

The second update on `Connection` is just an extension of the `Socket\Connection` to use the specific Socket extension... But we lose the registered wrappers here... Maybe we can make a mix.

Hywan replied:
Or, the factory, instead of being declared as a static method on `Hoa\Websocket\Connection` can (must?) be declared as a static method on `Hoa\Websocket\Socket` and return a `new static('tcp://…')`. It makes more sense! Thoughts?

shulard replied:
Yes it's more clean to let the `Websocket\Socket` object decide how to handle himself... I'll update!

shulard commented:
Hello @hoaproject/hoackers !

I worked on the socket wrapper implementation in Websocket library. There is a code design issue here which make the code work depending on instruction sequence...

I gave exemple [here](https://github.com/hoaproject/Websocket/pull/47#issue-98134425)...

Any though about that ?
Thanks a lot for your help, I can't find a clean way to deal with it :smile:

Hywan commented:
@shulard Did you see my comment here: https://github.com/hoaproject/Websocket/pull/47#discussion_r35869872?

shulard commented:
Yes and I've updated the code with that method :smile:

We discussed about that here [#47 (comment)](https://github.com/hoaproject/Websocket/pull/47#discussion_r35952109) and decided to move the factory in Websocket\Socket (done in 1eb1730).

Hywan commented:
So, what's next?

shulard commented:
In fact I don't know... I think that the current implementation is not really nice regarding to the instruction sequence that can be broke...

This can be documented but it's not cool for users to have to deal with that.

Maybe we can add a factory to build the `Hoa\Websocket\Client`. Then the wrappers can be registered at the right time...

```php
$client = Hoa\Websocket\Client::fromUri("wss://0.0.0.0/a/path");
```

Hywan commented:
The current usage:
```php
$socket = new Hoa\Socket\Client('ws://…');
$websocket = new Hoa\Websocket\Client($socket);
```

However, the `ws://` and `wss://` wrappers need to be registered. This is done by the `Hoa\Websocket` library. And thus, it must be loaded before. This is what you are mentionning?

So… either the user makes sure the `ws://` or `ws://` wrappers are registered, or `Hoa\Websocket\Client` and `Hoa\Websocket\Server` can accept a string instead of an `Hoa\Socket\Connection` instance.

Thoughts?

shulard commented:
It's exactly what I'm mentioning. 

I think that let the user detect if the wrappers are registered isn't the right thing to do. Wrappers are hidden inside the `Websocket` lib and not exposed by default.

So I'm going to update the `Websocket\Client` and `Websocket\Server` constructors to accept strings and take care of object creation.

Hywan commented:
@shulard Let me know when a review will be required.

shulard commented:
I think it's ok now... I've updated constructors behaviour + documentation. Now you can pass a string or a valid instance to build a `Websocket\Client` or `Websocket\Server` instance.

Hywan commented:
We should do that for all `Hoa\Socket` children…

What about addressing this issue in another PR?

shulard commented:
Of course I can make a specific PR :smile:

Indeed, this constructor update does not rely specifically with the wrapper handling system. I'll linked this PR to the new one and rollback the specific commits here.

Hywan commented:
Thanks.

shulard committed:
Add "files" autoload to always load wrappers.

shulard commented:
Composer behaviour updated to handle wrapper autoloading by default... This update was discussed here: https://github.com/hoaproject/Websocket/pull/50#issuecomment-143165558

It seems that this update must be in this PR instead of #50...

I think that this PR is ready for final review now...

Hywan commented:
So, it is ready for review and merge? Because we have 3 or maybe 4 PR, we have to do this in the correct order. Can you come on IRC to discuss about it (if you think this is required, or if it can replaced by a short comment)?

Hywan commented:
@shulard Could you make a summary in this meta issue please https://github.com/hoaproject/Websocket/issues/44?

shulard commented:
I've done the summary of the different PRs in #44.

shulard committed:
Handle endPoint in a better way

If null is given we rely on the socket endpoint if available. We always set the endPoint on the current Client instead.

shulard commented:
I've rebased commit here too :smile:

Hywan commented on Client.php:
```diff
@@ -83,11 +83,20 @@ class Client extends Connection
+                $endPoint = $client->getSocket()->getEndPoint();
+            }
+
+        }
```
CS: Add a newline after `}` please.

Hywan commented on Server.php:
```diff
@@ -67,7 +67,7 @@ class Server extends Connection
     public function __construct(
-        Socket\Server $server,
+        HoaSocket\Server $server,
         Http\Request  $request = null
```
CS: Columnize please (add 3 spaces before `$request).

Hywan commented on Socket.php:
```diff
@@ -0,0 +1,135 @@
+     *
+     * @var boolean
+     */
+    protected $_secure   = false;
```
Not necessary, already in `Hoa\Socket\Socket` (https://github.com/hoaproject/Socket/blob/2f0ad39705018fed38b7350ccc5d4f3dfbee6573/Socket.php#L112) under `_secured`.

Hywan commented on Socket.php:
```diff
@@ -0,0 +1,135 @@
+    {
+        parent::__construct($uri);
+
+        $this->_secure = $secure;
```
```diff
- $this->_secure
+ $this->_secured
```

Hywan commented on Socket.php:
```diff
@@ -0,0 +1,135 @@
+     *
+     * @return boolean
+     */
+    public function isSecure()
```
Not necessary, already in `Hoa\Socket\Socket` (https://github.com/hoaproject/Socket/blob/2f0ad39705018fed38b7350ccc5d4f3dfbee6573/Socket.php#L112) under `_secured`.

Hywan commented on Socket.php:
```diff
@@ -0,0 +1,135 @@
+
+    /**
+     * Factory to create a valid instance from URI
+     * @param string $socketUri
```
Documentation: Please, complete the API documentation.

Hywan commented on Socket.php:
```diff
@@ -0,0 +1,135 @@
+    public static function createFromUri($socketUri)
+    {
+        $parsed = parse_url($socketUri);
+        if( false === $parsed ) {
```
CS: Please, use `hoa devtools:cs`.

Hywan commented on Socket.php:
```diff
@@ -0,0 +1,135 @@
+        $parsed = parse_url($socketUri);
+        if( false === $parsed ) {
+            throw new Exception(
+                'URI %s is not recognized.',
```
Please, can you provide a better (more explicit) exception message?

Hywan commented on composer.json:
```diff
@@ -32,7 +32,8 @@
             "Hoa\\Websocket\\": "."
-        }
+        },
+        "files": [ "Connection.php" ]
```
What about loading `Socket.php` instead? Make more sense.

shulard replied:
Yes it make more sens :+1:

Hywan commented on Connection.php:
```diff
@@ -555,3 +555,9 @@ public function close($code = self::CLOSE_NORMAL, $reason = null)
+/**
+ * Register socket wrappers
+ */
+HoaSocket\Transport::registerWrapper('ws', ['Hoa\Websocket\Socket', 'createFromUri']);
```
Please, move this in `Socket.php`, see https://github.com/hoaproject/Websocket/pull/47/files#r46777215 for more explanations.

Hywan commented on Socket.php:
```diff
@@ -0,0 +1,135 @@
+     * @param string $socketUri
+     * @return void
+     */
+    public static function createFromUri($socketUri)
```
Please, rename to `transportFactory`. What do you think? This is more consistent with recent patches on `Hoa\Socket` (https://github.com/hoaproject/Socket/pull/27#issuecomment-162353542).

shulard committed:
Move wrapper registration to Socket extension

shulard committed:
Use the new Hoa\Socket API

shulard committed:
Fix CS

shulard commented:
@Hywan, I've updated the code with your comments and with the Socket API update.

shulard commented:
Hello !
Any news about that PR ? Now the related Socket PR is merged so I think we can go forward on this one :smile:.

Do you want that I rebase the 3 last commits into 1 ?

Hywan commented:
Not yet. I am going to review them, and after that we will rebase if necessary.

Hywan commented:
Hello :-),

I have rebased your patches. I have also changed the commit messages.
Here is the result:
  * [Socket: Introduce the `Socket` class](https://github.com/hoaproject/Websocket/commit/53712fd79db6f4f64f343bb61dc727328fd10fcf),
  * [Socket: Introduce `ws://` and `wss://` transports](https://github.com/hoaproject/Websocket/commit/c9f3f92192e3c8b2215ba0e8901a234824b541b3),
  * [Quality: Fix CS](https://github.com/hoaproject/Websocket/commit/759479c23ca48e7bb02b0ed6aed58ad5883962cd).

Then, I have added the following commits:
  * [Composer: Require `hoa/test`](https://github.com/hoaproject/Websocket/commit/b591016afe8cfa96677f1fb31f85d231b6c08dbe),
  * [Socket: Detect invalid URI in transport factory](https://github.com/hoaproject/Websocket/commit/41d7c8a60bbdd993ca2db5a19c8a856eef100be0),
  * [Test: Write test suite of `Hoa\Websocket\Socket`](https://github.com/hoaproject/Websocket/commit/eb45c00be5d55eb3c35f80fd1ccc04d3cd0d33f5).

The test suite contains:
  * 12 test cases,
  * 106 assertions.

The test suite will fail until https://github.com/hoaproject/Socket/pull/35 is merged.

Hywan commented:
Thanks!

shulard commented:
Hello @Hywan,

Thanks for your detailed review! A lot of inspiration for me :smile:.

Happy that this feature is now merged \o/

Hywan commented:
@shulard Do you know think you learn something that is absent from the contributor guide and should be present?

shulard commented:
@Hywan, I think it's more about the testing logic than the contribution process... I'll take a deeper look at contributor guide and submit PR if I have some ideas :smile:.
